### PR TITLE
Differentiate code challenges and ranks

### PR DESCRIPTION
**Pull request by SquishyStrawberry**
This branch fixes the issue of different endpoints having different formats.

Please review the challenges and comment with any problems, or comment just to say you're okay with the merge.

### CODE REVIEW

**Comment by theelous3** [comment]
:yes:

**Comment by SquishyStrawberry** [comment]
:thumbsup:

File: pythonwars/util.py [hunk 1]
```diff
@@ -12,6 +14,15 @@ def __init__(self, json_data):
...
+            else:
+                class_name = None
             for inner_key in tree:
                 look_in = look_in[inner_key]
-            setattr(self, inflection.underscore(leaf), look_in[leaf])
+            result = look_in[leaf]
+            if class_name is not None:
+                *module_name, class_name = class_name.split(".")
+                module_name = ".".join(module_name)
+                result = sys.modules[module_name].__dict__[class_name](result)
```
Comment (raylu) [inline pythonwars/util.py:27]: this is an interesting way of writing `getattr`

File: pythonwars/codechallenge.py [hunk 2]
```diff
@@ -17,3 +18,23 @@ class CodeChallenge(AttributesFromJson):
...
+    attr_keys = [
+        "id",
+        "name",
+        "slug",
+        "category",
+        "publishedAt",
+        "approvedAt",
+        "languages",
+        "url",
+        "rank:pythonwars.user.CodeChallengeRank",
```
Comment (raylu) [inline pythonwars/codechallenge.py:32]: maybe this should be `('rank', 'pythonw...')`

File: pythonwars/pythonwars.py [hunk 3]
```diff
@@ -42,13 +42,13 @@ def train_next_code_challenge(self, language, strategy="default", peek=False):
...
         response.raise_for_status()
-        return CodeChallenge(response.json())
+        return TrainingCodeChallenge(response.json())
 
     def train_code_challenge(self, id_or_slug, language):
         url = _TRAIN_CODE_CHALLENGE_URL.format(id_or_slug, language)
         response = self.session.post(url)
         response.raise_for_status()
-        return CodeChallenge(response.json())
+        return TrainingCodeChallenge(response.json())
```
Comment (raylu) [inline pythonwars/pythonwars.py:51]: this is starting to look familiar... a class method on `AttributesFromJson`, perhaps? (also should rename it at that point)

File: pythonwars/util.py (hunk 1 above)
Comment (Akuli) [inline pythonwars/util.py:27]: This is just crazy. `sys.modules` and attribute tricks (`__dict__` or `getattr`) just for a very basic attribute holder class?

**Pull request merged**